### PR TITLE
[FIRRTL] Dedup memory wrapper modules in LowerMemory

### LINKED ISSUE #6445: [Dedup][FIRRTL] SRAM modules not deduped when using memory macro replacement
# Context

Platform: macOS 14.1.1
Architecture: arm64
circt version: Firtool 1.58.0 release, also f124fb041

# Description

When generating a design from Chisel with arrays of SRAM instances, I notice that the SRAM instances are placed inside of modules that aren't deduped. Here's a small Chisel design to reproduce the issue with an array of SRAMs, each with a single read-write port:

```scala
class Foo extends Module {
  val numMems    = 2
  val numEntries = 16
  val dType      = UInt(8.W)

  val addr   = IO(Vec(numMems, Input(UInt(log2Ceil(numEntries).W))))
  val inData = IO(Vec(numMems, Input(dType)))
  val out    = IO(Vec(numMems, Output(dType)))
  val wen    = IO(Vec(numMems, Input(Bool())))

  for (i <- 0 until numMems) {
    val mem = SyncReadMem(numEntries, dType)

    val rwPort = mem(addr(i))
    out(i) := DontCare
    when (wen(i)) {
      rwPort := inData(i)
    }.otherwise {
      out(i) := rwPort
    }
  }
}
```

If you run that through a Chisel 3.6.0 flow, using the SFC...

```scala
//> using scala "2.13.10"
//> using lib "edu.berkeley.cs::chisel3::3.6.0"
//> using plugin "edu.berkeley.cs:::chisel3-plugin::3.6.0"
//> using options "-unchecked", "-deprecation", "-language:reflectiveCalls", "-feature", "-Xcheckinit", "-Xfatal-warnings", "-Ywarn-dead-code", "-Ywarn-unused", "-Ymacro-annotations"

import chisel3._
import chisel3.util.log2Ceil

class Foo extends Module {
  // Insert design from above
}

object Main extends App {
  val useFirtool = false
  if (!useFirtool) {
    val pretty = Array(
      "--infer-rw",
      "--repl-seq-mem",
      "-c:Foo:-o:foo.conf",
      "--emission-options",
      "disableMemRandomization,disableRegisterRandomization",
    )
    println(getVerilogString(new Foo, pretty))
  } else {
    println(
      circt.stage.ChiselStage.emitSystemVerilog(
        gen = new Foo,
        firtoolOpts = Array(
          "-disable-all-randomization",
          "-strip-debug-info",
          "-repl-seq-mem",
          "-repl-seq-mem-file=foo.conf",
        ),
      )
    )
  }
}
```

Then you get this:

```verilog
module Foo(
  input        clock,
  input        reset,
  input  [3:0] addr_0,
  input  [3:0] addr_1,
  input  [7:0] inData_0,
  input  [7:0] inData_1,
  output [7:0] out_0,
  output [7:0] out_1,
  input        wen_0,
  input        wen_1
);
  wire [3:0] mem_RW0_addr;
  wire  mem_RW0_clk;
  wire  mem_RW0_wmode;
  wire [7:0] mem_RW0_wdata;
  wire [7:0] mem_RW0_rdata;
  wire [3:0] mem_1_RW0_addr;
  wire  mem_1_RW0_clk;
  wire  mem_1_RW0_wmode;
  wire [7:0] mem_1_RW0_wdata;
  wire [7:0] mem_1_RW0_rdata;
  mem mem (
    .RW0_addr(mem_RW0_addr),
    .RW0_clk(mem_RW0_clk),
    .RW0_wmode(mem_RW0_wmode),
    .RW0_wdata(mem_RW0_wdata),
    .RW0_rdata(mem_RW0_rdata)
  );
  mem mem_1 (
    .RW0_addr(mem_1_RW0_addr),
    .RW0_clk(mem_1_RW0_clk),
    .RW0_wmode(mem_1_RW0_wmode),
    .RW0_wdata(mem_1_RW0_wdata),
    .RW0_rdata(mem_1_RW0_rdata)
  );
  assign out_0 = mem_RW0_rdata;
  assign out_1 = mem_1_RW0_rdata;
  assign mem_RW0_addr = addr_0;
  assign mem_RW0_clk = clock;
  assign mem_RW0_wmode = wen_0;
  assign mem_RW0_wdata = inData_0;
  assign mem_1_RW0_addr = addr_1;
  assign mem_1_RW0_clk = clock;
  assign mem_1_RW0_wmode = wen_1;
  assign mem_1_RW0_wdata = inData_1;
endmodule
module mem(
  input  [3:0] RW0_addr,
  input        RW0_clk,
  input        RW0_wmode,
  input  [7:0] RW0_wdata,
  output [7:0] RW0_rdata
);
  wire [3:0] mem_ext_RW0_addr;
  wire  mem_ext_RW0_en;
  wire  mem_ext_RW0_clk;
  wire  mem_ext_RW0_wmode;
  wire [7:0] mem_ext_RW0_wdata;
  wire [7:0] mem_ext_RW0_rdata;
  mem_ext mem_ext (
    .RW0_addr(mem_ext_RW0_addr),
    .RW0_en(mem_ext_RW0_en),
    .RW0_clk(mem_ext_RW0_clk),
    .RW0_wmode(mem_ext_RW0_wmode),
    .RW0_wdata(mem_ext_RW0_wdata),
    .RW0_rdata(mem_ext_RW0_rdata)
  );
  assign mem_ext_RW0_clk = RW0_clk;
  assign mem_ext_RW0_en = 1'h1;
  assign mem_ext_RW0_addr = RW0_addr;
  assign RW0_rdata = mem_ext_RW0_rdata;
  assign mem_ext_RW0_wmode = RW0_wmode;
  assign mem_ext_RW0_wdata = RW0_wdata;
endmodule
```

Notice that there's only one `mem` module definition, which wraps around `mem_ext`.

If you run this through firtool (change that `useFirtool` flag from `false` to `true`), you get this:

```verilog
// Generated by CIRCT unknown git version
module Foo(
  input        clock,
               reset,
  input  [3:0] addr_0,
               addr_1,
  input  [7:0] inData_0,
               inData_1,
  output [7:0] out_0,
               out_1,
  input        wen_0,
               wen_1
);

  mem mem (
    .RW0_addr  (addr_0),
    .RW0_clk   (clock),
    .RW0_wmode (wen_0),
    .RW0_wdata (inData_0),
    .RW0_rdata (out_0)
  );
  mem_1 mem_1 (
    .RW0_addr  (addr_1),
    .RW0_clk   (clock),
    .RW0_wmode (wen_1),
    .RW0_wdata (inData_1),
    .RW0_rdata (out_1)
  );
endmodule

module mem(
  input  [3:0] RW0_addr,
  input        RW0_clk,
               RW0_wmode,
  input  [7:0] RW0_wdata,
  output [7:0] RW0_rdata
);

  mem_ext mem_ext (
    .RW0_addr  (RW0_addr),
    .RW0_en    (1'h1),
    .RW0_clk   (RW0_clk),
    .RW0_wmode (RW0_wmode),
    .RW0_wdata (RW0_wdata),
    .RW0_rdata (RW0_rdata)
  );
endmodule

// external module mem_ext

module mem_1(
  input  [3:0] RW0_addr,
  input        RW0_clk,
               RW0_wmode,
  input  [7:0] RW0_wdata,
  output [7:0] RW0_rdata
);

  mem_ext mem_ext (
    .RW0_addr  (RW0_addr),
    .RW0_en    (1'h1),
    .RW0_clk   (RW0_clk),
    .RW0_wmode (RW0_wmode),
    .RW0_wdata (RW0_wdata),
    .RW0_rdata (RW0_rdata)
  );
endmodule


// ----- 8< ----- FILE "metadata/seq_mems.json" ----- 8< -----

[
  {
    "module_name": "mem_ext",
    "depth": 16,
    "width": 8,
    "masked": false,
    "read": 0,
    "write": 0,
    "readwrite": 1,
    "extra_ports": [],
    "hierarchy": [
      "Foo.mem_1.mem_ext",
      "Foo.mem.mem_ext"
    ]
  }
]

// ----- 8< ----- FILE "foo.conf" ----- 8< -----

name mem_ext depth 16 width 8 ports rw
```

While there's only one `mem_ext` definition, the wrapper modules `mem`/`mem_1` aren't deduplicated.

### PR DESCRIPTION
**Pull request by tymcauley**
Instead of just dedup-ing the external memory module, we include the memory wrapper module in the dedup calculation.

Resolves #6445.

Diff in `circt-opt -firrtl-lower-memory test/Dialect/FIRRTL/lower-memory.mlir` after this change:

```diff
--- main
+++ dedup-lower-memory-modules
@@ -75,7 +75,7 @@
   firrtl.circuit "Dedup" {
     firrtl.module @Dedup() {
       %mem0_W0_addr, %mem0_W0_en, %mem0_W0_clk, %mem0_W0_data = firrtl.instance mem0 @mem0(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
-      %mem1_W0_addr, %mem1_W0_en, %mem1_W0_clk, %mem1_W0_data = firrtl.instance mem1 @mem1(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
+      %mem1_W0_addr, %mem1_W0_en, %mem1_W0_clk, %mem1_W0_data = firrtl.instance mem1 @mem0(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
     }
     firrtl.module private @mem0(in %W0_addr: !firrtl.uint<4>, in %W0_en: !firrtl.uint<1>, in %W0_clk: !firrtl.clock, in %W0_data: !firrtl.uint<42>) {
       %mem0_ext_W0_addr, %mem0_ext_W0_en, %mem0_ext_W0_clk, %mem0_ext_W0_data = firrtl.instance mem0_ext @mem0_ext(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
@@ -85,13 +85,6 @@
       firrtl.strictconnect %mem0_ext_W0_data, %W0_data : !firrtl.uint<42>
     }
     firrtl.memmodule private @mem0_ext(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>) attributes {dataWidth = 42 : ui32, depth = 12 : ui64, extraPorts = [], maskBits = 1 : ui32, numReadPorts = 0 : ui32, numReadWritePorts = 0 : ui32, numWritePorts = 1 : ui32, readLatency = 1 : ui32, writeLatency = 1 : ui32}
-    firrtl.module private @mem1(in %W0_addr: !firrtl.uint<4>, in %W0_en: !firrtl.uint<1>, in %W0_clk: !firrtl.clock, in %W0_data: !firrtl.uint<42>) {
-      %mem0_ext_W0_addr, %mem0_ext_W0_en, %mem0_ext_W0_clk, %mem0_ext_W0_data = firrtl.instance mem0_ext @mem0_ext(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
-      firrtl.strictconnect %mem0_ext_W0_addr, %W0_addr : !firrtl.uint<4>
-      firrtl.strictconnect %mem0_ext_W0_en, %W0_en : !firrtl.uint<1>
-      firrtl.strictconnect %mem0_ext_W0_clk, %W0_clk : !firrtl.clock
-      firrtl.strictconnect %mem0_ext_W0_data, %W0_data : !firrtl.uint<42>
-    }
   }
   firrtl.circuit "NoTestharnessDedup0" {
     firrtl.module @NoTestharnessDedup0() {
@@ -257,30 +250,23 @@
     hw.hierpath private @nla0 [@NonLocalAnnotation::@dut, @DUT::@mem0]
     hw.hierpath private @nla0_0 [@NonLocalAnnotation::@dut, @DUT::@mem0, @mem0]
     hw.hierpath private @nla1 [@NonLocalAnnotation::@dut, @DUT]
-    hw.hierpath private @nla1_0 [@NonLocalAnnotation::@dut, @DUT::@sym, @mem1]
+    hw.hierpath private @nla1_0 [@NonLocalAnnotation::@dut, @DUT::@sym, @mem0]
     firrtl.module @NonLocalAnnotation() {
       firrtl.instance dut sym @dut @DUT()
     }
     firrtl.module @DUT() {
       %mem0_W0_addr, %mem0_W0_en, %mem0_W0_clk, %mem0_W0_data = firrtl.instance mem0 sym @mem0 @mem0(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
-      %mem1_W0_addr, %mem1_W0_en, %mem1_W0_clk, %mem1_W0_data = firrtl.instance mem1 sym @sym @mem1(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
+      %mem1_W0_addr, %mem1_W0_en, %mem1_W0_clk, %mem1_W0_data = firrtl.instance mem1 sym @sym @mem0(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
       %MRead_read = firrtl.mem Undefined {depth = 12 : i64, name = "MRead", portNames = ["read"], readLatency = 0 : i32, writeLatency = 1 : i32} : !firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data flip: uint<42>>
     }
     firrtl.module private @mem0(in %W0_addr: !firrtl.uint<4>, in %W0_en: !firrtl.uint<1>, in %W0_clk: !firrtl.clock, in %W0_data: !firrtl.uint<42>) {
-      %mem0_ext_W0_addr, %mem0_ext_W0_en, %mem0_ext_W0_clk, %mem0_ext_W0_data = firrtl.instance mem0_ext sym @mem0_ext {annotations = [{circt.nonlocal = @nla0_0, class = "test0"}]} @mem0_ext(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
+      %mem0_ext_W0_addr, %mem0_ext_W0_en, %mem0_ext_W0_clk, %mem0_ext_W0_data = firrtl.instance mem0_ext sym @mem0_ext {annotations = [{circt.nonlocal = @nla0_0, class = "test0"}, {circt.nonlocal = @nla1_0, class = "test1"}]} @mem0_ext(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
       firrtl.strictconnect %mem0_ext_W0_addr, %W0_addr : !firrtl.uint<4>
       firrtl.strictconnect %mem0_ext_W0_en, %W0_en : !firrtl.uint<1>
       firrtl.strictconnect %mem0_ext_W0_clk, %W0_clk : !firrtl.clock
       firrtl.strictconnect %mem0_ext_W0_data, %W0_data : !firrtl.uint<42>
     }
     firrtl.memmodule private @mem0_ext(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>) attributes {dataWidth = 42 : ui32, depth = 12 : ui64, extraPorts = [], maskBits = 1 : ui32, numReadPorts = 0 : ui32, numReadWritePorts = 0 : ui32, numWritePorts = 1 : ui32, readLatency = 1 : ui32, writeLatency = 1 : ui32}
-    firrtl.module private @mem1(in %W0_addr: !firrtl.uint<4>, in %W0_en: !firrtl.uint<1>, in %W0_clk: !firrtl.clock, in %W0_data: !firrtl.uint<42>) {
-      %mem0_ext_W0_addr, %mem0_ext_W0_en, %mem0_ext_W0_clk, %mem0_ext_W0_data = firrtl.instance mem0_ext sym @mem0_ext {annotations = [{circt.nonlocal = @nla1_0, class = "test1"}]} @mem0_ext(in W0_addr: !firrtl.uint<4>, in W0_en: !firrtl.uint<1>, in W0_clk: !firrtl.clock, in W0_data: !firrtl.uint<42>)
-      firrtl.strictconnect %mem0_ext_W0_addr, %W0_addr : !firrtl.uint<4>
-      firrtl.strictconnect %mem0_ext_W0_en, %W0_en : !firrtl.uint<1>
-      firrtl.strictconnect %mem0_ext_W0_clk, %W0_clk : !firrtl.clock
-      firrtl.strictconnect %mem0_ext_W0_data, %W0_data : !firrtl.uint<42>
-    }
   }
 }
```

### CODE REVIEW

File: lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp:96]: Using tuples tends to make code larger and less clear than making a struct.

File: lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp
Comment (darthscsi) [reply lib/Dialect/FIRRTL/Transforms/LowerMemory.cpp:276]: The instance name is not guaranteed to be unique with respect to inner symbol names.

**Review by darthscsi — COMMENTED** [review]
In the PR comment, I don't understand `@mem0` instantiating mem0_ext_0 and mem0_ext.  Shouldn't `@Dedup` simply be instantiating mem0 twice?  The change to the test in the PR looks more like what I'd expect.

**Comment by tymcauley** [comment]
> In the PR comment, I don't understand `@mem0` instantiating mem0_ext_0 and mem0_ext. Shouldn't `@Dedup` simply be instantiating mem0 twice? The change to the test in the PR looks more like what I'd expect.

Yup, you're right. There's a bug in my change right now, I need to move the `memInst` creation to be at the point where we define the `wrapper` module. I _think_ this should also explain the issue I was having with the duplicate inner symbols.

Thanks for pointing that out!

**Comment by tymcauley** [comment]
Side question: should the `@nla0` and `@nla1` non-local annotations be deleted during this pass, or are they cleaned up elsewhere, or should we just leave them be?

**Comment by darthscsi** [comment]
> Side question: in the `NonLocalAnnotation` circuit in `lower-memory.mlir`, should the `@nla0` and `@nla1` non-local annotations be deleted during this pass (since they refer to the original `firrtl.mem` instances), or are they cleaned up elsewhere, or should we just leave them be?

If you change then hierarchy, you need to update the hierpaths.  Dedup is somewhat unique in that it has to interpret the NLAs to decide if they matter or not (e.g. block dedup).  (this is part of what makes dedup fragile)

**Comment by prithayan** [comment]
Sorry for the delay, I will take a look at this today.

**Review by prithayan — APPROVED** [review]
The changes make sense. Dead `HierPathOp` can be left for cleanup by later passes, but if symbol names are updated the `HierPathOp` needs to be updated and cannot be left in an inconsistent state. The pass is already handling them correctly.

**Comment by tymcauley** [comment]
> The changes make sense. Dead `HierPathOp` can be left for cleanup by later passes, but if symbol names are updated the `HierPathOp` needs to be updated and cannot be left in an inconsistent state. The pass is already handling them correctly.

Thanks for clearing that up @prithayan! Is there anything I should take care of before this can be merged? Let me know if you'd like me to rebase on the latest `main`.

**Comment by prithayan** [comment]
> > The changes make sense. Dead `HierPathOp` can be left for cleanup by later passes, but if symbol names are updated the `HierPathOp` needs to be updated and cannot be left in an inconsistent state. The pass is already handling them correctly.
> 
> Thanks for clearing that up @prithayan! Is there anything I should take care of before this can be merged? Let me know if you'd like me to rebase on the latest `main`.

Please rebase and merge it.

**Comment by tymcauley** [comment]
> > > The changes make sense. Dead `HierPathOp` can be left for cleanup by later passes, but if symbol names are updated the `HierPathOp` needs to be updated and cannot be left in an inconsistent state. The pass is already handling them correctly.
> > 
> > 
> > Thanks for clearing that up @prithayan! Is there anything I should take care of before this can be merged? Let me know if you'd like me to rebase on the latest `main`.
> 
> Please rebase and merge it.

I've rebased, but I don't have write permissions, so someone else will have to merge it.

**Comment by tymcauley** [comment]
@prithayan just rebased again, can you merge this please? Thanks!

**Comment by prithayan** [comment]
Sorry, missed to merge it last week. 
Thanks for fixing the issue :) 
Looking forward to more future contribution @tymcauley .

**Comment by tymcauley** [comment]
> Sorry, missed to merge it last week. Thanks for fixing the issue :) Looking forward to more future contribution @tymcauley .

No problem @prithayan, thanks so much for taking care of this!

**Pull request merged**
